vulkan/vulkan_swapchain.cc

```diff
@@ -0,0 +1,511 @@
+      .fImageLayout = VK_IMAGE_LAYOUT_UNDEFINED,
+      .fFormat = surface_format_.format,
+      .fLevelCount = 1,
+  };
+
+  GrBackendRenderTargetDesc desc;
+  desc.fWidth = size.fWidth;
+  desc.fHeight = size.fHeight;
+  desc.fConfig = pixel_config;
+  desc.fOrigin = kTopLeft_GrSurfaceOrigin;
+  desc.fSampleCnt = 0;
+  desc.fStencilBits = 0;
+  desc.fRenderTargetHandle = reinterpret_cast<GrBackendObject>(&image_info);
+
+  SkSurfaceProps props(SkSurfaceProps::InitType::kLegacyFontHost_InitType);
+
+  return SkSurface::MakeFromBackendRenderTarget(gr_context, desc, &props);
+}
+
+bool VulkanSwapchain::CreateSwapchainImages(GrContext* skia_context) {
+  std::vector<VkImage> images = GetImages();
+
+  if (images.size() == 0) {
+    return false;
+  }
+
+  const SkISize surface_size = GetSize();
+
+  for (const VkImage& image : images) {
+    // Populate the backbuffer.
+    auto backbuffer = std::make_unique<VulkanBackbuffer>(vk, device_.Handle(),
+                                                         device_.CommandPool());
+
+    if (!backbuffer->IsValid()) {
+      return false;
+    }
+
+    backbuffers_.emplace_back(std::move(backbuffer));
+
+    // Populate the image.
+    auto vulkan_image = std::make_unique<VulkanImage>(image);
+
+    if (!vulkan_image->IsValid()) {
+      return false;
+    }
+
+    images_.emplace_back(std::move(vulkan_image));
+
+    // Populate the surface.
+    auto surface = CreateSkiaSurface(skia_context, image, surface_size);
+
+    if (surface == nullptr) {
+      return false;
+    }
+
+    surfaces_.emplace_back(std::move(surface));
+  }
+
+  FTL_DCHECK(backbuffers_.size() == images_.size());
+  FTL_DCHECK(images_.size() == surfaces_.size());
+
+  return true;
+}
+
+VulkanBackbuffer* VulkanSwapchain::NextBackbuffer() {
+  auto available_backbuffers = backbuffers_.size();
+
+  if (available_backbuffers == 0) {
+    return nullptr;
+  }
+
+  auto next_backbuffer_index =
+      (current_backbuffer_index_ + 1) % backbuffers_.size();
+
+  auto& backbuffer = backbuffers_[next_backbuffer_index];
+
+  if (!backbuffer->IsValid()) {
+    return nullptr;
+  }
+
+  current_backbuffer_index_ = next_backbuffer_index;
+  return backbuffer.get();
+}
+
+VulkanSwapchain::AcquireResult VulkanSwapchain::AcquireSurface() {
+  AcquireResult error = {AcquireType::ErrorSurfaceLost, nullptr};
+
+  if (!IsValid()) {
+    return error;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Step 0:
+  // Acquire the next available backbuffer.
+  // ---------------------------------------------------------------------------
+  auto backbuffer = NextBackbuffer();
+
+  if (backbuffer == nullptr) {
+    return error;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Step 1:
+  // Wait for use readiness.
+  // ---------------------------------------------------------------------------
+  if (!backbuffer->WaitFences()) {
```

Comment:
It shouldn't be necessary to stall the CPU waiting for a fence.  Instead, you can use the semaphore passed to AcquireNextImageKHR(), and wait for it when submitting the command-buffer that first renders into it.

For example: https://github.com/fuchsia-mirror/escher/blob/master/escher/vk/vulkan_swapchain_helper.cc#L47

Perhaps the problem is that Skia doesn't expose an API to pass it the semaphore?